### DISCUSSION TITLE
Maneiras de gerar Critical CSS

### DISCUSSION THREAD

**Discussion by woliveiras** [frontendbr/forum#381] [Discussão]
Fala galera, estou abrindo essa issue para perguntar como vocês estão gerando o Critical CSS de suas aplicações.

Aqui temos uma tarefa no Grunt que roda a task para gerar e adicionar o Critical na hora do deploy. Porém para isso acontecer a aplicação precisa estar rodando via localhost:8000, onde a tarefa vai analisar o código e retornar somente o que aparece acima da borda.

A dúvida é: teria alguma maneira de gerar isso sem a necessidade da aplicação estar rodando?

Nós pensamos em diversas maneiras, mas eu gostaria de validar com vocês como estão fazendo para coletar mais dicas. 😄

**Comment by jeffersondanielss** [comment]
Cara eu estava discutindo isso a alguns dias atrás com um colega de trabalho, dei uma pesquisada em algumas ferramentas para isso, ainda não testei mas acho vale dar uma conferida [nessa](https://github.com/addyosmani/critical).

**Comment by gbrlsepulveda** [comment]
Fala @woliveiras, boa sugestão de tema.
Utilizo essa ferramenta que o @jeffersondanielss citou, o processo de build ficou relativamente complexo. Durante o processo de build, levantamos a aplicação e rodamos a task de critical apontando para o arquivo que deve ser extraido o CSS. Funcionava muito bem, até um momento que passou a estourar um erro de uma dependência do projeto. Ainda não conseguimos contornar essa situação, estamos testando outras ferramentas

**Comment by fdaciuk** [comment]
Cheguei em algumas soluções interessantes com webpack.. o exemplo está [nesse projeto](https://github.com/da2k/curso-reactjs-ninja/blob/master/examples/m02/working-with-assets/github-app), quando é gerado o build para produção.

A ideia basicamente é:
- Usar o plugin `extract-text-webpack-plugin` para gerar [um arquivo para o CRP](https://github.com/da2k/curso-reactjs-ninja/blob/master/examples/m02/working-with-assets/github-app/webpack.prod.config.js#L64-L67) e outro para o [resto dos estilos](https://github.com/da2k/curso-reactjs-ninja/blob/master/examples/m02/working-with-assets/github-app/webpack.prod.config.js#L59-L62);

Nesse caso, eu setei manualmente os arquivos que eu queria que fossem usados para o CRP, mas como o webpack permite usar regex para escolher os arquivos, daria pra colocar no final do arquivo um `.crp.css`.. os que tiverem essa extensão vão para o arquivo `crp.css` no final;

- Estou usando o `html-webpack-plugin` para gerar o [HTML final](https://github.com/da2k/curso-reactjs-ninja/blob/master/examples/m02/working-with-assets/github-app/webpack.prod.config.js#L38-L42);

- Dentro do [HTML gerado](https://github.com/da2k/curso-reactjs-ninja/blob/master/examples/m02/working-with-assets/github-app/src/html/template.html), eu pego o conteúdo do `crp.css` gerado e coloco inline no `<head>`, e os outros estilos e scripts no footer da aplicação =)

Com isso já resolve uma parte do problema =)

O próximo passo é fazer isso usando `react-router` e com `server-rendering`, que estou testando em outro projeto. Provavelmente vou gravar uns vídeos falando sobre o assunto, e como eu cheguei nessa solução xD

Com certeza não é a melhor solução, mas resolve bem o problema, permitindo ainda componentizar a aplicação de forma bem fácil =)

**Comment by woliveiras** [comment]
Que loucura galera.

Fazemos bem parecido com seu caso @gbrlsepulveda.

E é essa a treta: queríamos parar de ter que subir a aplicação. :grin:

Talvez um esquema parecido com o que o @fdaciuk comentou e setar manualmente seja o caminho pra o caso lá no trampo.

Vamos testar!!

**Comment by tarciozemel** [comment]
Passando só para lembrar que [não se deve usar CSS crítico em todo e qualquer caso](http://desenvolvimentoparaweb.com/css/css-critico-acima-da-borda-desmistificando/).
